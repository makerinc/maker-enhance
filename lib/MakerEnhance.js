"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = MakerEnhance;
var _react = _interopRequireDefault(require("react"));
var _MakerEnhanceClient = _interopRequireDefault(require("./MakerEnhanceClient"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function MakerEnhance(_ref) {
  var user = _ref.user,
    index = _ref.index,
    loadingHeight = _ref.loadingHeight;
  var scriptSrc = "https://app.maker.co/enhance/".concat(user, ".js");
  var html = "\n    <script src=\"".concat(scriptSrc, "\" id=\"maker-enhance-script\" async=\"true\"></script>\n    <div\n      id=\"js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-").concat(index || 0, "\"\n      class=\"js-maker-enhance-static-mount\"\n      style=\"height:auto;width:100%\"\n      ").concat(loadingHeight ? "data-loading-height=\"".concat(loadingHeight, "\"") : "", "\n    ></div>\n  ");

  // https://github.com/facebook/react/issues/10923#issuecomment-338715787
  // "We won't try to manipulate the tree of a dangerouslySetInnerHTML node on the client. Even if it is wrong."
  //
  // Use dangerouslySetInnerHTML to avoid issues with hydration.
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_MakerEnhanceClient["default"], {
    user: user,
    index: index,
    scriptSrc: scriptSrc
  }), /*#__PURE__*/_react["default"].createElement("div", {
    className: "js-maker-enhance-wrapper",
    dangerouslySetInnerHTML: {
      __html: html
    },
    suppressHydrationWarning: true
  }));
}