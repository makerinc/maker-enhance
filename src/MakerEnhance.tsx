import React from "react";
import MakerEnhanceClient from "./MakerEnhanceClient";

interface MakerEnhanceProps {
  user: string;
  index?: number;
  loadingHeight?: string;
}

export default function MakerEnhance({
  user,
  index,
  loadingHeight
}: MakerEnhanceProps): JSX.Element {
  const scriptSrc = `https://app.maker.co/enhance/${user}.js`;

  const html = `
    <script src="${scriptSrc}" id="maker-enhance-script" async="true"></script>
    <div
      id="js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-${index ||
        0}"
      class="js-maker-enhance-static-mount"
      style="height:auto;width:100%"
      ${loadingHeight ? `data-loading-height="${loadingHeight}"` : ""}
    ></div>
  `;

  // https://github.com/facebook/react/issues/10923#issuecomment-338715787
  // "We won't try to manipulate the tree of a dangerouslySetInnerHTML node on the client. Even if it is wrong."
  //
  // Use dangerouslySetInnerHTML to avoid issues with hydration.
  return (
    <>
      <MakerEnhanceClient user={user} index={index} scriptSrc={scriptSrc} />
      <div
        className="js-maker-enhance-wrapper"
        dangerouslySetInnerHTML={{ __html: html }}
        suppressHydrationWarning={true}
      />
    </>
  );
}
