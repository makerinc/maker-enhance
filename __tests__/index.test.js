import React from "react";
import renderer from "react-test-renderer";
import MakerEnhance from "../src/MakerEnhance";

let mocks = {};

beforeEach(() => {
  document.head.innerHTML = "";
});

afterEach(() => {
  Object.keys(mocks).forEach(key => {
    mocks[key].mockRestore();
  });

  mocks = {};
});

test("MakerEnhance renders div with default index", () => {
  const component = renderer.create(<MakerEnhance user="linkesch" />);
  const tree = component.toJSON();
  const scripts = document.querySelectorAll("#maker-enhance-script");

  expect(tree).toMatchSnapshot();
  expect(scripts.length).toBe(1);
  expect(scripts[0].src).toBe("https://app.maker.co/enhance/linkesch.js");
});

test("MakerEnhance renders div with provided index", () => {
  const component = renderer.create(<MakerEnhance user="linkesch" index="2" />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test("MakerEnhance should not add script if user prop is not provided", () => {
  mocks = {
    error: jest.spyOn(window.console, "error").mockImplementation(() => null)
  };

  const component = renderer.create(<MakerEnhance />);
  const tree = component.toJSON();
  const scripts = document.head.querySelectorAll("#maker-enhance-script");

  expect(tree).toMatchSnapshot();
  expect(scripts.length).toBe(0);
});

test("MakerEnhance should call run() function on props update", () => {
  window.MakerEmbeds = {
    run: jest.fn()
  };

  const component = renderer.create(<MakerEnhance user="linkesch" />);
  component.update(<MakerEnhance user="linkesch" index={1} />);

  expect(window.MakerEmbeds.run).toHaveBeenCalled();
});

test("MakerEnhance should call run() function on update when also url changes", () => {
  window.MakerEmbeds = {
    run: jest.fn()
  };

  const component = renderer.create(<MakerEnhance user="linkesch" />);
  window.location.href = "#test";

  component.update(<MakerEnhance user="linkesch" />);

  expect(window.MakerEmbeds.run).toHaveBeenCalled();
});

test("MakerEnhance should not call run() function on useless update", () => {
  window.MakerEmbeds = {
    run: jest.fn()
  };

  const component = renderer.create(<MakerEnhance user="linkesch" />);
  component.update(<MakerEnhance user="linkesch" />);

  expect(window.MakerEmbeds.run).toHaveBeenCalledTimes(1);
});

test("MakerEnhance should support loadingHeight prop", () => {
  const component = renderer.create(
    <MakerEnhance user="linkesch" loadingHeight={100} />
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
