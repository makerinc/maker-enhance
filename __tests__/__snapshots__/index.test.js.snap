// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MakerEnhance renders div with default index 1`] = `
<div
  className="js-maker-enhance-wrapper"
  dangerouslySetInnerHTML={
    Object {
      "__html": "
      <script src=\\"https://app.maker.co/enhance/linkesch.js\\" id=\\"maker-enhance-script\\" async=\\"true\\"></script>
      <div
        id=\\"js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-0\\"
        class=\\"js-maker-enhance-static-mount\\"
        style=\\"height:auto;width:100%\\"
        
      ></div>
    ",
    }
  }
  suppressHydrationWarning={true}
/>
`;

exports[`MakerEnhance renders div with provided index 1`] = `
<div
  className="js-maker-enhance-wrapper"
  dangerouslySetInnerHTML={
    Object {
      "__html": "
      <script src=\\"https://app.maker.co/enhance/linkesch.js\\" id=\\"maker-enhance-script\\" async=\\"true\\"></script>
      <div
        id=\\"js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-2\\"
        class=\\"js-maker-enhance-static-mount\\"
        style=\\"height:auto;width:100%\\"
        
      ></div>
    ",
    }
  }
  suppressHydrationWarning={true}
/>
`;

exports[`MakerEnhance should not add script if user prop is not provided 1`] = `
<div
  className="js-maker-enhance-wrapper"
  dangerouslySetInnerHTML={
    Object {
      "__html": "
      <script src=\\"https://app.maker.co/enhance/undefined.js\\" id=\\"maker-enhance-script\\" async=\\"true\\"></script>
      <div
        id=\\"js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-0\\"
        class=\\"js-maker-enhance-static-mount\\"
        style=\\"height:auto;width:100%\\"
        
      ></div>
    ",
    }
  }
  suppressHydrationWarning={true}
/>
`;

exports[`MakerEnhance should support loadingHeight prop 1`] = `
<div
  className="js-maker-enhance-wrapper"
  dangerouslySetInnerHTML={
    Object {
      "__html": "
      <script src=\\"https://app.maker.co/enhance/linkesch.js\\" id=\\"maker-enhance-script\\" async=\\"true\\"></script>
      <div
        id=\\"js-maker-static-enhance-v1-218c2d7d-62da-499e-9e74-93201e1b3d56-0\\"
        class=\\"js-maker-enhance-static-mount\\"
        style=\\"height:auto;width:100%\\"
        data-loading-height=\\"100\\"
      ></div>
    ",
    }
  }
  suppressHydrationWarning={true}
/>
`;
